### PR TITLE
Add version to cache key for chat and llm blocks

### PR DESCRIPTION
**Pull request by flvndvd**
## Description

This PR aims to improve the release process for structural changes and prevent regressions similar to the one introduced in #4895. The regression in #4895 missed an issue with the cache. For more details, please refer to the full context provided [here](https://dust4ai.slack.com/archives/C050SM8NSPK/p1714461084925719).

We updated the `ChatFunctionCall` to include an `id` field. However, this new field was not marked as optional, leading to the inability to deserialize cached data from the cache table to the updated structure.

To prevent this issue from recurring, this PR:
- Introduces a new trait called `CachedRequest` that must be implemented for every request to be cached. This trait includes:
  - A version that should be incremented whenever the types undergo changes.
  - A type that indicates the request type, which will be used to easily invalidate old rows.
- Updates the cache table to add two new columns: `version` and `type` to ease removing legacy rows.

The version is included in the hash and will generate new cache objects automatically if altered. This will enable us to easily increase the version and ensure a seamless rollout if modifications are needed for the chat or LLM block.

## Risk

When release all the cached objects will be outdated. We will be able to delete all rows from the `cache` table that don't contain any type and version. This PR remains safe to rollback. 

## Deploy Plan

This PR requires a migration to add the new `type` and `version` field on the `cache` table.

The following needs to be run:
```
ALTER TABLE cache
ADD COLUMN version INTEGER,
ADD COLUMN type TEXT;
```

### CODE REVIEW

**Review by spolu — APPROVED** [review]
LGTM

Deploy should include deleting the whole cache

File: core/src/providers/llm.rs [hunk 1]
```diff
@@ -336,6 +336,7 @@ pub struct LLMChatRequest {
     presence_penalty: Option<f32>,
     frequency_penalty: Option<f32>,
     extras: Option<Value>,
+    version: i32,
```
Comment (spolu) [inline core/src/providers/llm.rs:339]: maybe cache_version?

**Review by spolu — COMMENTED** [review]
Maybe we should insert in the cache table the type and cache_version so that we can clean up properly as we bump versions. But this is not urgent as we can simply delete everything this time around and the next few ones.

**Comment by flvndvd** [comment]
> Maybe we should insert in the cache table the type and cache_version so that we can clean up properly as we bump versions. But this is not urgent as we can simply delete everything this time around and the next few ones.

Yep I thought about adding the block type at least. I'll bake this change in this PR and add the right index so we can delete from the cache table.

File: core/src/providers/llm.rs (hunk 1 above)
Comment (flvndvd) [reply core/src/providers/llm.rs:339]: 🤔 It's not the cache version but more the block's version actually.

Comment (spolu) [reply core/src/providers/llm.rs]: Yeah well, it's meant to be used for the cache only so might as well be explicit. No strong fell anyway. It's just that `version` alone might raise some questions to the reader that are bigger than this really is.

**Comment by flvndvd** [comment]
> > Maybe we should insert in the cache table the type and cache_version so that we can clean up properly as we bump versions. But this is not urgent as we can simply delete everything this time around and the next few ones.
> 
> Yep I thought about adding the block type at least. I'll bake this change in this PR and add the right index so we can delete from the cache table.

@spolu PR updated using a trait. PR's description updated as well to reflect the latest changes. I'll add the necessary scripts to 1. delete the legacy rows and 2. delete rows for an older version in a follow up PR.

**Review by spolu — APPROVED** [review]
Looks great.

**Comment by spolu** [comment]
Exact deploy is
- add tables to core
- deploy `core`

Will test it locally and merge

**Comment by spolu** [comment]
Checked that the alter table is safe to run before deploying on my devbox

**Comment by spolu** [comment]
Fixed the migration that said request_type when we really wanted type :)

**Comment by spolu** [comment]
Deploying works well, cache is operational.

**Comment by spolu** [comment]
```
dust-api=> SELECT COUNT(*)F FROM cache WHERE version IS NULL;
    f    
---------
 2948931
(1 row)
```

**Comment by spolu** [comment]
Nuking them all
```
DELETE FROM cache WHERE version IS NULL;
```

**Comment by flvndvd** [comment]
Thank you for shpping 🔥 !

**Pull request merged**